File: react-app/src/store/venue.js
const READ_SINGLE_VENUE = 'venue/READ_SINGLE_VENUE'
const READ_ALL_VENUES = 'venue/READ_ALL_VENUES'
const CREATE_VENUE = 'venue/CREATE_VENUE'
const UPDATE_VENUE = 'venue/UPDATE_VENUE'
const DELETE_VENUE = 'venue/DELETE_VENUE'


const loadOneVenue = venue => ({
    type: READ_SINGLE_VENUE,
    payload: venue
});

const loadAllVenues = venues => ({
    type: READ_ALL_VENUES,
    payload: venues
});

const addVenue = venue => ({
    type: CREATE_VENUE,
    payload: venue
});

const changeVenue = venue => ({
    type: UPDATE_VENUE,
    payload: venue
});

const removeVenue = venue => ({
    type: DELETE_VENUE,
    payload: venue
});


export const getSingleVenue = venueId => async dispatch => {
    const res = await fetch(`/api/venues/${venueId}/`)
    const data = await res.json();

    if (res.ok) {
        dispatch(loadOneVenue(data))
    } else {
        throw res;
    };
}

export const getAllVenues = () => async dispatch => {
    const res = await fetch(`/api/venues/`);
    const data = await res.json();

    if (res.ok) {
        dispatch(loadAllVenues(data))
    } else {
        throw res;
    };
}

export const getAllUserVenues = userId => async dispatch => {
    const res = await fetch(`/api/venues/users/${userId}/`);
    const data = await res.json();

    if (res.ok) {
        dispatch(loadAllVenues(data))
    } else {
        throw res;
    };
}

export const createVenue = venueData => async dispatch => {
    const { userId, name, dateTypeId, price, description, imageUrl, phoneNumber, address, city, state, zipcode, operationHours, latitude, longitude } = venueData
    const res = await fetch(`/api/venues/create/`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            creator_id: userId,
            name: name,
            date_type_id: dateTypeId,
            price: price,
            description: description,
            image_url: imageUrl,
            phone_number: phoneNumber,
            address: address,
            city: city,
            state: state,
            zipcode: zipcode,
            operation_hours: operationHours,
            latitude: latitude,
            longitude: longitude
        }),
    });
    const data = await res.json();

    if (res.ok) {
        dispatch(addVenue(data))
    } else {
        throw res;
    };
    return data
}

export const updateVenue = venueData => async dispatch => {
    const { userId, name, dateTypeId, price, description, imageUrl, phoneNumber, address, city, state, zipcode, operationHours, latitude, longitude, venueId } = venueData
    const res = await fetch(`/api/venues/${venueId}/`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            creator_id: userId,
            name: name,
            date_type_id: dateTypeId,
            price: price,
            description: description,
            image_url: imageUrl,
            phone_number: phoneNumber,
            address: address,
            city: city,
            state: state,
            zipcode: zipcode,
            operation_hours: operationHours,
            latitude: latitude,
            longitude: longitude
        }),
    });
    const data = await res.json();

    if (res.ok) {
        dispatch(changeVenue(data))
    } else {
        throw res;
    };
    return data
}

export const deleteVenue = venueId => async dispatch => {
    const res = await fetch(`/api/venues/${venueId}/`, {
        method: 'DELETE',
    })

    if (res.ok) {
        dispatch(removeVenue(venueId))
    } else {
        throw res;
    };
}


const initialState = {}
const venuesReducer = (state = initialState, action) => {
    let newState;
    switch (action.type) {
        case READ_SINGLE_VENUE:
            newState = { ...state };
            newState.current = action.payload;
            return newState;

        case READ_ALL_VENUES:
            newState = { ...action.payload };
            return newState;

        case CREATE_VENUE:
            newState = Object.assign({}, state);
            newState[action.payload.id] = action.payload;
            return newState;

        case UPDATE_VENUE:
            newState = { ...state };
            newState.current = action.payload;
            return newState;

        case DELETE_VENUE:
            newState = { ...state };
            delete newState[action.payload];
            return newState;

        default:
            return state;
    }
}

export default venuesReducer;

File: app/models/venue.py
from .db import db
from flask_login import UserMixin


class Venue(db.Model, UserMixin):
    __tablename__ = 'venues'

    id = db.Column(db.Integer, primary_key=True)
    creator_id = db.Column(db.Integer, db.ForeignKey('users.id'))
    name = db.Column(db.String(40), nullable=False, unique=True)
    date_type_id = db.Column(db.Integer, db.ForeignKey(
        'date_types.id'), nullable=False)
    price = db.Column(db.Integer, nullable=False)
    description = db.Column(db.Text, nullable=False)
    image_url = db.Column(db.String(255))
    phone_number = db.Column(db.String(255))
    address = db.Column(db.String(255), nullable=False)
    city = db.Column(db.String(100), nullable=False)
    state = db.Column(db.String(25), nullable=False)
    zipcode = db.Column(db.String(10), nullable=False)
    operation_hours = db.Column(db.String(255), nullable=False)
    latitude = db.Column(db.Float, nullable=False)
    longitude = db.Column(db.Float, nullable=False)

    user = db.relationship('User', back_populates='venues')
    date_type = db.relationship('Date_type', back_populates='venues')
    reviews = db.relationship('Review', back_populates='venue')
    reservations = db.relationship('Reservation', back_populates='venue')
    favorites = db.relationship("Favorite", back_populates="venue")

    def to_dict(self):
        return {
            'id': self.id,
            'creator_id': self.creator_id,
            'username': self.name,
            'name': self.name,
            'price': self.price,
            'date_type_id': self.date_type_id,
            'description': self.description,
            'image_url': self.image_url,
            'phone_number': self.phone_number,
            'address': self.address,
            'city': self.city,
            'state': self.state,
            'zipcode': self.zipcode,
            'operation_hours': self.operation_hours,
            "latitude": self.latitude,
            "longitude": self.longitude,
            "reviews": {review.id: review.to_dict() for review in self.reviews},
            "favorites": {favorite.id: favorite.to_dict() for favorite in self.favorites}
        }
